File: src/main/java/com/derongan/minecraft/guiy/gui/layouts/HistoryGuiHolder.java
package com.derongan.minecraft.guiy.gui.layouts;

import com.derongan.minecraft.guiy.gui.*;
import de.erethon.headlib.HeadLib;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;

import java.util.ArrayList;
import java.util.List;

public class HistoryGuiHolder extends GuiHolder {
    protected Player player;
    private List<Layout> history = new ArrayList<>();
    private ClickableElement back;

    public HistoryGuiHolder(int numRows, String title, Plugin plugin) {
        super(numRows, title, new Layout(), plugin);

        //create back button
        Element cell = Cell.forItemStack(HeadLib.RED_X.toItemStack("Back"));
        back = new ClickableElement(cell);
        back.setClickAction(clickEvent -> backInHistory());
    }

    public ClickableElement getBackButton() {
        return back;
    }

    public void addBackButton(Layout layout) {
        history.add(layout);
        layout.addElement(8, 5, back);
    }

    public void backInHistory() {
        if (history.size() <= 1 && player != null) {
            player.closeInventory();
            return;
        }
        Layout previous = history.get(history.size() - 2);
        history.remove(history.size() - 1);

        setElement(previous);
    }
}

File: src/main/java/com/derongan/minecraft/guiy/gui/layouts/HistoryGuiHolder.kt
package com.derongan.minecraft.guiy.gui.layouts

import com.derongan.minecraft.guiy.gui.Element
import com.derongan.minecraft.guiy.gui.GuiHolder
import com.derongan.minecraft.guiy.gui.Layout
import com.derongan.minecraft.guiy.gui.elements.ClickableElement
import com.derongan.minecraft.guiy.helpers.toCell
import de.erethon.headlib.HeadLib
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin
import java.util.*

@RequiresOptIn(level = RequiresOptIn.Level.WARNING, message = "This is an experimental feature in Guiy designed for Kotlin. It may be changed in the future without notice.")
annotation class GuiyExperimentalKotlinAPI

@GuiyExperimentalKotlinAPI
abstract class HistoryGuiHolder @JvmOverloads constructor(numRows: Int, title: String, plugin: Plugin, initial: Element = Layout()) : GuiHolder(numRows, title, plugin, initial) {
    private var player: Player? = null
    private val history: MutableList<Layout> = ArrayList()
    private val backButton: ClickableElement

    private fun addBackButton(layout: Layout) {
        history.add(layout)
        layout.setElement(8, 5, backButton)
    }

    fun backInHistory() {
        if (history.size <= 1 && player != null) {
            player!!.closeInventory()
            return
        }
        //TODO change to removeLast when that's no longer marked experimental
        history.removeAt(history.size - 1)
        setElement(history.last())
        render()
    }

    override fun show(player: Player) {
        setElement(root)
        super.show(player)
        this.player = player
    }

    /** Sets the element, then adds a back button if it's a layout*/
    override fun setElement(element: Element) {
        super.setElement(element)
        if (element is Layout && history.lastOrNull() != element)
            addBackButton(element)
    }

    init {
        //create back button
        val cell = HeadLib.RED_X.toCell("Back")
        backButton = ClickableElement(cell) { backInHistory() }
    }
}